designs/v1beta1-full-changelist.md

```diff
@@ -0,0 +1,200 @@
+      requirements: ...
+      providerRef: ...
+  deprovisioning:
+    ttlUntilExpired: ...
+    ttlAfterUnderutilized: ...
+```
+
+#### `provisioner.Spec.ttl...` → `nodePool.Spec.deprovisioning.ttl...`
+
+Karpenter plans to expand the amount of control that it gives users over both the aggressiveness of deprovisioning and when deprovisioning can take place. As part of these upcoming changes, more fields within the `NodePool` API will begint to pertain to deprovisioning configuration.
+
+We can better delineate the fields that specifically pertain to this configuration from the other fields in the `spec` (global behavior-based fields, provisioning-specific fields, node static configuration fields) by moving these fields inside a `deprovisioning` block. This will make it clearer to users which configuration options specifically pertain to scale-down when they are configuring their `NodePool` CRs.
+
+#### `provisioner.Spec.ttlSecondsAfterEmpty` → `nodePool.spec.deprovisioning.consolidationPolicy`
+
+Currently, Karpenter has two mutually exclusive ways to deprovision nodes based on emptiness: `ttlSecondsAfterEmpty` and `consolidation`. If users are using `ttlSecondsAfterEmpty`, we have generally seen that users are configuring this field in one of two ways:
+
+1. `ttlSecondsAfterEmpty=0` → Users want to delete nodes as soon as they go empty and Karpenter sees that they are empty
+2. `ttlSecondsAfterEmpty >> 0` → Users want to delete nodes that are empty but want to reduce the amount of node churn as a result of high pod churn on a larger cluster
+
+We anticipate that both of these scenarios can be captured through the consolidation deprovisioning mechanism; however, we understand that there are use-cases where a user may want to reduce the aggressiveness of Karpenter deprovisioning and only deprovision empty nodes. In this case, a user can configure the `consolidationPolicy` to be `WhenEmpty` which will tell the consolidation deprovisioning mechanism to only deprovision empty nodes through consolidation. Alternatively, you can specify a `consolidationPolicy` of `WhenUnderutilized` which will allow consolidation to deprovision both empty and underutilized nodes.
+
+If `consolidationPolicy` is not set, Karpenter will implicitly default to `WhenUnderutilized`.
+
+#### `provisioner.Spec.ttlSecondsAfterEmpty` → `nodePool.spec.deprovisioning.consolidationTTL`
+
+While the `consolidationPolicy` offers one mechanism for users to control the aggressiveness of deprovisioning, users that enable a `consolidationPolicy` of `WhenEmpty` or `WhenUnderutilized` may still want to dictate the speed at which nodes are deemed underutilized. This is particularly true on clusters that are large in size and have a large amount of churn. To support this, Karpenter will surface a `consolidationTTL` which will allow you to define a per-node TTL to define the time that Karpenter can begin deprovisioning the node after first seeing that the node is eligible for consolidation.
```

Comment:
Depending on how much we like CEL, we could allow people to specify a _consolidation eligibility expression_.
(for early beta, that could be further constrained to say something like

“OK, but _for now_ the only supported expressions look like `node.secondsSinceEmpty >= *`, more later”

Reply:
> we could allow people to specify a __consolidation eligibility expression__

I'm curious what you're thinking here. Like define their own custom selector logic based on creationTimestamp or the lastTransitionTime of a condition field?

Reply:
Not a selector, but broadly yes. However this is mostly an aside / tangent.

Reply:
Expanding on this, it would be cool if eventually deprovisioning for all 5 types of deprovisioning can be controlled. Such as turning on and off each type of deprovisioning and the TTL.